File: apps/web/src/lib/bbcode/__tests__/bbcode.test.ts
/**
 * BBCode Parser Tests
 *
 * Tests for BBCode parsing, tag handling, stripping, validation,
 * and sanitization helpers.
 */

import { describe, it, expect } from 'vitest';
import {
  parseBBCode,
  stripBBCode,
  countBBCodeCharacters,
  validateBBCode,
  escapeHtml,
  extractYouTubeId,
  isValidColor,
  sanitizeFontSize,
  sanitizeFontFamily,
} from '../index';

// ---------------------------------------------------------------------------
// parseBBCode
// ---------------------------------------------------------------------------

describe('parseBBCode', () => {
  it('returns empty string for empty input', () => {
    expect(parseBBCode('')).toBe('');
  });

  it('parses [b] bold tags', () => {
    const result = parseBBCode('[b]hello[/b]');
    expect(result).toContain('<strong');
    expect(result).toContain('hello');
  });

  it('parses [i] italic tags', () => {
    const result = parseBBCode('[i]italic[/i]');
    expect(result).toContain('<em');
    expect(result).toContain('italic');
  });

  it('parses [u] underline tags', () => {
    const result = parseBBCode('[u]underline[/u]');
    expect(result).toContain('underline');
    expect(result).toContain('class="underline"');
  });

  it('parses [s] strikethrough tags', () => {
    const result = parseBBCode('[s]deleted[/s]');
    expect(result).toContain('<del');
    expect(result).toContain('deleted');
  });

  it('parses nested formatting tags', () => {
    const result = parseBBCode('[b][i]bold italic[/i][/b]');
    expect(result).toContain('<strong');
    expect(result).toContain('<em');
  });

  it('parses [code] blocks and escapes HTML inside', () => {
    const result = parseBBCode('[code]<script>alert(1)</script>[/code]');
    expect(result).toContain('<pre');
    expect(result).toContain('<code');
    expect(result).not.toContain('<script>alert(1)</script>');
  });

  it('parses [quote] blocks', () => {
    const result = parseBBCode('[quote]some quote[/quote]');
    expect(result).toContain('<blockquote');
  });

  it('parses [quote=Author] blocks with author attribution', () => {
    const result = parseBBCode('[quote=John]words[/quote]');
    expect(result).toContain('John');
    expect(result).toContain('wrote:');
  });

  it('parses [spoiler] tags into <details>', () => {
    const result = parseBBCode('[spoiler]hidden[/spoiler]');
    expect(result).toContain('<details');
    expect(result).toContain('hidden');
  });

  it('converts newlines to <br />', () => {
    const result = parseBBCode('line1\nline2');
    expect(result).toContain('<br />');
  });

  it('parses [color=red] tags', () => {
    const result = parseBBCode('[color=red]colored[/color]');
    expect(result).toContain('color: red');
  });

  it('parses [hr] as a horizontal rule', () => {
    const result = parseBBCode('[hr]');
    expect(result).toContain('<hr');
  });

  it('handles [list] with items', () => {
    const result = parseBBCode('[list][*]one[*]two[/list]');
    expect(result).toContain('<ul');
    expect(result).toContain('<li');
  });
});

// ---------------------------------------------------------------------------
// stripBBCode
// ---------------------------------------------------------------------------

describe('stripBBCode', () => {
  it('returns empty string for empty input', () => {
    expect(stripBBCode('')).toBe('');
  });

  it('strips bold tags and keeps text', () => {
    expect(stripBBCode('[b]hello[/b]')).toBe('hello');
  });

  it('strips url tags and keeps the display text', () => {
    expect(stripBBCode('[url=https://example.com]Click here[/url]')).toBe('Click here');
  });

  it('strips [img] tags and their content', () => {
    // [img] content is replaced with [Image], then the cleanup regex strips bracket-wrapped text
    expect(stripBBCode('[img]https://example.com/photo.png[/img]')).toBe('');
  });

  it('replaces bullet points with •', () => {
    const result = stripBBCode('[list][*]item[/list]');
    expect(result).toContain('•');
  });
});

// ---------------------------------------------------------------------------
// countBBCodeCharacters
// ---------------------------------------------------------------------------

describe('countBBCodeCharacters', () => {
  it('counts only the visible text, not tags', () => {
    expect(countBBCodeCharacters('[b]hello[/b]')).toBe(5);
  });

  it('returns 0 for empty input', () => {
    expect(countBBCodeCharacters('')).toBe(0);
  });
});

// ---------------------------------------------------------------------------
// validateBBCode
// ---------------------------------------------------------------------------

describe('validateBBCode', () => {
  it('returns valid for properly closed tags', () => {
    const result = validateBBCode('[b]text[/b]');
    expect(result.valid).toBe(true);
    expect(result.errors).toHaveLength(0);
  });

  it('detects unclosed [b] tag', () => {
    const result = validateBBCode('[b]text');
    expect(result.valid).toBe(false);
    expect(result.errors[0]).toContain('[b]');
  });

  it('detects extra closing tag without opener', () => {
    const result = validateBBCode('text[/i]');
    expect(result.valid).toBe(false);
    expect(result.errors[0]).toContain('[/i]');
  });

  it('validates complex nested content as valid', () => {
    const result = validateBBCode('[b][i]nested[/i][/b]');
    expect(result.valid).toBe(true);
  });
});

// ---------------------------------------------------------------------------
// BBCode sanitizer helpers
// ---------------------------------------------------------------------------

describe('escapeHtml (bbcode)', () => {
  it('escapes angle brackets', () => {
    expect(escapeHtml('<div>')).toContain('&lt;');
    expect(escapeHtml('<div>')).toContain('&gt;');
  });

  it('escapes ampersands', () => {
    expect(escapeHtml('a & b')).toContain('&amp;');
  });
});

describe('extractYouTubeId', () => {
  it('extracts ID from standard watch URL', () => {
    expect(extractYouTubeId('https://www.youtube.com/watch?v=dQw4w9WgXcQ')).toBe('dQw4w9WgXcQ');
  });

  it('extracts ID from short URL', () => {
    expect(extractYouTubeId('https://youtu.be/dQw4w9WgXcQ')).toBe('dQw4w9WgXcQ');
  });

  it('extracts ID from embed URL', () => {
    expect(extractYouTubeId('https://www.youtube.com/embed/dQw4w9WgXcQ')).toBe('dQw4w9WgXcQ');
  });

  it('extracts ID from shorts URL', () => {
    expect(extractYouTubeId('https://youtube.com/shorts/dQw4w9WgXcQ')).toBe('dQw4w9WgXcQ');
  });

  it('returns null for non-YouTube URL', () => {
    expect(extractYouTubeId('https://vimeo.com/12345')).toBeNull();
  });
});

describe('isValidColor', () => {
  it('accepts hex colors', () => {
    expect(isValidColor('#ff0000')).toBe(true);
    expect(isValidColor('#abc')).toBe(true);
  });

  it('accepts named colors', () => {
    expect(isValidColor('red')).toBe(true);
    expect(isValidColor('Blue')).toBe(true);
  });

  it('rejects invalid color strings', () => {
    expect(isValidColor('notacolor')).toBe(false);
    expect(isValidColor('#xyz')).toBe(false);
  });
});

describe('sanitizeFontSize', () => {
  it('clamps numeric sizes between 8 and 36', () => {
    expect(sanitizeFontSize('4')).toBe('8px');
    expect(sanitizeFontSize('50')).toBe('36px');
    expect(sanitizeFontSize('16')).toBe('16px');
  });

  it('maps named sizes', () => {
    expect(sanitizeFontSize('small')).toBe('14px');
    expect(sanitizeFontSize('large')).toBe('18px');
  });

  it('defaults to 16px for unknown named sizes', () => {
    expect(sanitizeFontSize('huge')).toBe('16px');
  });
});

describe('sanitizeFontFamily', () => {
  it('allows whitelisted fonts', () => {
    expect(sanitizeFontFamily('Arial')).toContain('Arial');
  });

  it('falls back to Arial for unknown fonts', () => {
    expect(sanitizeFontFamily('EvilFont')).toBe('"Arial", sans-serif');
  });
});

File: apps/backend/test/cgraph/forums/bbcode_test.exs
defmodule CGraph.Forums.BBCodeTest do
  use ExUnit.Case, async: true

  alias CGraph.Forums.BBCode

  describe "to_html/1" do
    test "returns empty string for nil" do
      assert BBCode.to_html(nil) == ""
    end

    test "escapes HTML entities before processing" do
      assert BBCode.to_html("<script>alert('xss')</script>") =~
               "&lt;script&gt;"
    end

    test "converts newlines to <br>" do
      assert BBCode.to_html("line1\nline2") == "line1<br>line2"
    end
  end

  describe "simple formatting tags" do
    test "[b] bold" do
      assert BBCode.to_html("[b]text[/b]") == "<strong>text</strong>"
    end

    test "[i] italic" do
      assert BBCode.to_html("[i]text[/i]") == "<em>text</em>"
    end

    test "[u] underline" do
      assert BBCode.to_html("[u]text[/u]") == "<u>text</u>"
    end

    test "[s] strikethrough" do
      assert BBCode.to_html("[s]text[/s]") == "<s>text</s>"
    end

    test "nested tags" do
      result = BBCode.to_html("[b][i]bold italic[/i][/b]")
      assert result == "<strong><em>bold italic</em></strong>"
    end
  end

  describe "[url] tag" do
    test "bare url" do
      result = BBCode.to_html("[url]https://example.com[/url]")
      assert result =~ ~s(href="https://example.com")
      assert result =~ "rel=\"nofollow noopener\""
    end

    test "url with label" do
      result = BBCode.to_html("[url=https://example.com]Click here[/url]")
      assert result =~ ~s(href="https://example.com")
      assert result =~ "Click here</a>"
    end

    test "rejects javascript: URLs" do
      result = BBCode.to_html("[url=javascript:alert(1)]click[/url]")
      refute result =~ "<a"
      refute result =~ "javascript:"
    end

    test "rejects data: URLs" do
      result = BBCode.to_html("[url=data:text/html,<script>]click[/url]")
      refute result =~ "<a"
    end

    test "allows mailto: URLs" do
      result = BBCode.to_html("[url=mailto:user@example.com]email[/url]")
      assert result =~ "mailto:user@example.com"
    end
  end

  describe "[img] tag" do
    test "renders image with safe URL" do
      result = BBCode.to_html("[img]https://example.com/photo.jpg[/img]")
      assert result =~ "<img"
      assert result =~ ~s(src="https://example.com/photo.jpg")
      assert result =~ "loading=\"lazy\""
    end

    test "rejects javascript: in img src" do
      result = BBCode.to_html("[img]javascript:alert(1)[/img]")
      refute result =~ "<img"
    end
  end

  describe "[quote] tag" do
    test "plain quote" do
      result = BBCode.to_html("[quote]Some text[/quote]")
      assert result == "<blockquote>Some text</blockquote>"
    end

    test "quote with author" do
      result = BBCode.to_html("[quote=JohnDoe]Some text[/quote]")
      assert result =~ "<blockquote>"
      assert result =~ "JohnDoe wrote:"
      assert result =~ "Some text"
    end
  end

  describe "[code] tag" do
    test "renders code block" do
      result = BBCode.to_html("[code]def foo; end[/code]")
      assert result =~ "<pre><code>"
      assert result =~ "def foo; end"
    end

    test "does not convert newlines inside code" do
      result = BBCode.to_html("[code]line1\nline2[/code]")
      assert result =~ "line1\nline2"
      refute result =~ "line1<br>line2"
    end
  end

  describe "[list] tag" do
    test "renders unordered list" do
      input = "[list][*]Item 1[*]Item 2[*]Item 3[/list]"
      result = BBCode.to_html(input)
      assert result =~ "<ul>"
      assert result =~ "<li>Item 1</li>"
      assert result =~ "<li>Item 2</li>"
      assert result =~ "<li>Item 3</li>"
      assert result =~ "</ul>"
    end
  end

  describe "[color] tag" do
    test "valid CSS color name" do
      result = BBCode.to_html("[color=red]text[/color]")
      assert result =~ "color:red"
      assert result =~ "<span"
    end

    test "valid hex color" do
      result = BBCode.to_html("[color=#ff0000]text[/color]")
      assert result =~ "color:#ff0000"
    end

    test "rejects invalid color value" do
      result = BBCode.to_html("[color=expression(alert())]text[/color]")
      refute result =~ "<span"
    end
  end

  describe "[size] tag" do
    test "maps size 1-7 to em values" do
      result = BBCode.to_html("[size=1]tiny[/size]")
      assert result =~ "font-size:0.6em"

      result = BBCode.to_html("[size=7]huge[/size]")
      assert result =~ "font-size:2em"
    end

    test "clamps oversized value to 7" do
      result = BBCode.to_html("[size=99]big[/size]")
      assert result =~ "font-size:2em"
    end
  end

  describe "[center] tag" do
    test "renders centered text" do
      result = BBCode.to_html("[center]centered[/center]")
      assert result =~ "text-align:center"
    end
  end

  describe "[spoiler] tag" do
    test "renders spoiler with details/summary" do
      result = BBCode.to_html("[spoiler]hidden[/spoiler]")
      assert result =~ "<details>"
      assert result =~ "<summary>Spoiler</summary>"
      assert result =~ "hidden"
    end
  end

  describe "XSS prevention" do
    test "script tag injection" do
      result = BBCode.to_html("<script>alert('xss')</script>")
      refute result =~ "<script>"
      assert result =~ "&lt;script&gt;"
    end

    test "event handler injection in bbcode" do
      result = BBCode.to_html("[url=https://x.com\" onmouseover=\"alert(1)]click[/url]")
      refute result =~ "onmouseover"
    end

    test "nested XSS in img" do
      result = BBCode.to_html("[img]https://x.com/a.jpg\" onerror=\"alert(1)[/img]")
      # Quotes are escaped in attributes
      refute result =~ "onerror"
    end

    test "javascript: protocol is blocked" do
      assert BBCode.safe_url?("javascript:alert(1)") == false
      assert BBCode.safe_url?("JAVASCRIPT:alert(1)") == false
      assert BBCode.safe_url?("data:text/html,<script>") == false
    end

    test "valid protocols are allowed" do
      assert BBCode.safe_url?("http://example.com") == true
      assert BBCode.safe_url?("https://example.com") == true
      assert BBCode.safe_url?("mailto:user@example.com") == true
    end
  end

  describe "edge cases" do
    test "unclosed tags left as text" do
      result = BBCode.to_html("[b]no closing tag")
      assert result == "[b]no closing tag"
    end

    test "empty input" do
      assert BBCode.to_html("") == ""
    end

    test "plain text without tags" do
      assert BBCode.to_html("Hello world") == "Hello world"
    end

    test "case-insensitive tags" do
      assert BBCode.to_html("[B]bold[/B]") == "<strong>bold</strong>"
    end
  end
end
